Implement wxGrid's overlay selection

AliKet opened the pull request:
see #22694

Commit:
wxQt: override wxDCImpl::DoDrawPolyPolygon()

Commit:
Added convenience functions, RefreshRect() and RefreshBlock(), to wxGrid

The benefit of the former will become apparent when overlay selection and
label highlighting are added in the upcoming commits.

Commit:
Added wxGridSelection::MergeAdjacent{Blocks,Rects}() helper functions

These will used to implement overlay selection in the upcoming commits.

Commit:
Added wxGrid::GetSelectedRectangles() member function

This function returns the selected blocks (those visible only) as a vector of
rectangles to construct the selection shape: rectangle, polygon or poly-polygon.

Will be used to implement overlay selection in the upcoming commits.

Commit:
Added overlay selection capability to wxGrid

Modernize the wxGrid selection appearance by drawing a semi-transparent area
overlaid on the selected cells.

Notice that this feature will not be used if wxUSE_GRAPHICS_CONTEXT is set to 0
or DisableOverlaySelection() had been called.

Commit:
Refactor wxGrid{Row,Column}HeaderRendererDefault::DrawBorder() common code

No real changes yet.

Commit:
Added row/col highlight capability to wxGrid

Commit:
Added Is{Row,Col}LabelHighlighted() member functions to wxGrid

Will be used when label highlighting is added in the upcoming commits.

No real changes yet.

Commit:
Refresh the old highlighted row/col labels for a non-visible selection/current cell

Commit:
Fix rendering the overlay selection on external DC (via wxGrid::Render())

This commit also fixes drawing frozen borders if any.

Commit:
Update Grid::SelectionRange of GridTestCase

Test whether selecting two adjacent blocks results in only one block containing
both blocks. This should be the case since this commit caad24cf83 (Added
wxGridSelection::MergeAdjacent{Blocks,Rects}() helper functions)

AliKet commented:
@vadz I'm not sure if it's intentional or not, but I noticed that `wxGrid::Render()` doesn't render frozen borders (if any) at all, and I fixed this in [b77b7e4](https://github.com/wxWidgets/wxWidgets/pull/24561/commits/b77b7e4f3d0521081cff477ee89225dff7192197). Just let me know if I should change anything here, TIA

![Screenshot from 2024-05-27 20-31-48](https://github.com/wxWidgets/wxWidgets/assets/7704771/6189e9b2-b5b7-491a-9972-b69f6c442f60)

![Screenshot from 2024-05-27 20-31-04](https://github.com/wxWidgets/wxWidgets/assets/7704771/5229eacf-003f-4a95-9d4e-991bc3639773)

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -907,4 +943,455 @@ void wxGridSelection::MergeOrAddBlock(wxGridBlockCoordsVector& blocks,
+
+void wxGridSelection::PolyPolygon::CalcBoundingBox(const std::vector<wxPoint>& points)
+{
+    std::vector<wxPoint>::const_iterator cItr = points.begin(),
+                                         cEnd = points.end();
+    while ( cItr != cEnd )
+    {
+        const wxPoint& point = *cItr++;
```
Why not just
```suggestion
    for ( const auto& point : points )
    {
```
?

vadz commented on src/generic/gridsel.cpp:
```diff
@@ -907,4 +943,455 @@ void wxGridSelection::MergeOrAddBlock(wxGridBlockCoordsVector& blocks,
+                poly.push_back(p.first);
+            }
+
+            std::vector<wxPoint>::const_iterator cItr = poly.begin(),
+                                                 cEnd = poly.end();
+            while ( cItr != cEnd )
+            {
+                const wxPoint& pt = *cItr++;
```
This could be replaced by a range-for too AFAICS.

vadz commented on src/generic/grid.cpp:
```diff
@@ -58,6 +61,17 @@ const char wxGridNameStr[] = "grid";
 #include <ctype.h>
 
+// Drawing transparent selection can only be done via a graphics context,
+// or a DC that has a native support for drawing with alpha, like standard
+// DCs on OSX and GTK3.
```
Sorry, it's a pity that you've already done it, but I think there might have been a misunderstanding in one of the previous discussions: I never wanted to imply that you should support using `wxDC` when `wxUSE_G_C==0` in these ports. First of all, I don't even this it can be disabled in them and, second, even if it could, it's definitely not worth it because nobdoy would do it in practice.

So I'd just remove the port checks and use only `#if wxUSE_G_C` (although you might want to `#define wxHAS_TRANSPARENT_SELECTION` in this case and check for this one below to make these checks more readable).

vadz commented on src/generic/grid.cpp:
```diff
@@ -6935,6 +7149,116 @@ void wxGrid::DrawRowLabel( wxDC& dc, int row )
+
+    const wxColour colBg = GetSelectionBackground();
+
+#ifdef __WXMSW__
```
I'd appreciate a comment explaining why do we do this only under MSW as it doesn't seem obvious.

AliKet replied:
I wanted to make the selection outline more bold but after thinking more about it I think I'll revert this as it doesn't seem necessary.

Here is how it looks without this dirty hack:

In light mode:
![grid_light](https://github.com/wxWidgets/wxWidgets/assets/7704771/12f09e8e-baed-4f1f-81be-a12d9d1c2b70)

In dark mode:
![grid_dark](https://github.com/wxWidgets/wxWidgets/assets/7704771/d283ae7c-5efd-43d1-a629-000c347c04f6)

vadz reviewed (commented):
I haven't tried it yet but globally this looks good, thank you for implementing it!

I wonder if there is some better/more common term for this than "transparent selection", would anybody know how it is called in the existing programs? "Overlay selection" maybe?

In any case, whatever its final name, the new `wxGrid` function for enabling it should be documented.

Finally, ideally, we'd extract (some of?) the new functions in `wx/private/grid.h` in order to allow testing them as it's slightly worrisome that there is a lot of not really obvious new code here which is not covered by the unit tests...

Thanks again for your work!

AliKet commented:
> I wonder if there is some better/more common term for this than "transparent selection", would anybody know how it is called in the existing programs? "Overlay selection" maybe?
> 
So according to this [page](https://www.infragistics.com/help/winforms/wingrid-enabling-selection-overlay-for-selected-cells):
- I'll change `IsUsingTransparentSelection()` to  `IsUsingSelectionOverlay()` or simply `IsSelectionOverlay()` ?
- Remove `EnableTransparentSelection()` because **selection overlay** will be used by default if supported. But...
- If the environment variable `WXGRID_NO_SELECTION_OVERLAY` is set, then the old why to draw the selection will be used.

What do you think ?

Commit:
Rename DrawSelection() to DrawOverlaySelection()

Commit:
Update griddemo sample

Override DrawHighlighted() in CustomColumnHeaderRenderer for demonstration
purposes.

Commit:
Move PolyPolygon & PolyPolygonHelper to private header

No real change in behaviour or functionality.

Commit:
Move MergeAdjacent{Blocks,Rects}() to private header

No real change in behaviour or functionality.

AliKet commented on samples/grid/griddemo.cpp:
```diff
@@ -74,6 +74,16 @@ class CustomColumnHeaderRenderer : public wxGridColumnHeaderRenderer
+                                 int WXUNUSED(rowOrCol)) const override
+    {
+        dc.SetPen(*wxTRANSPARENT_PEN);
+        dc.SetBrush(wxBrush(wxColour(240, 200, 15, 180)));
```
The custom header renderer uses hardcoded colours and as such I don't think that's necessary here:

![Screenshot from 2024-05-30 15-08-41](https://github.com/wxWidgets/wxWidgets/assets/7704771/e913cb00-555c-46dc-bcc7-5689f7c0e870)

Commit:
wxQt: Added OffsetHelper class to wxGraphicsContext

vadz commented:
> > I wonder if there is some better/more common term for this than "transparent selection", would anybody know how it is called in the existing programs? "Overlay selection" maybe?
> 
> So according to this [page](https://www.infragistics.com/help/winforms/wingrid-enabling-selection-overlay-for-selected-cells):
> 
>     * I'll change `IsUsingTransparentSelection()` to  `IsUsingSelectionOverlay()` or simply `IsSelectionOverlay()` ?

I'd invert the word order as it seems more natural to me (can any native English speakers please comment on whether it's just me?), i.e. would use `IsUsingOverlaySelection()` or `UsesOverlaySelection()` or maybe `HasOverlaySelection()`, but any of these choices are acceptable, so please choose whichever you prefer.

>     * Remove `EnableTransparentSelection()` because **selection overlay** will be used by default if supported. But...
>     * If the environment variable `WXGRID_NO_SELECTION_OVERLAY` is set, then the old why to draw the selection will be used.

I don't see why should we have an environment variable to control something like this, so I'd strongly prefer to avoid it. But we could (should?) provide `DisableOverlaySelection()` to revert to the old appearance for people who would want it, for whatever reason.

Thanks again!

AliKet commented:
> I'd invert the word order as it seems more natural to me (can any native English speakers please comment on whether it's just me?), i.e. would use `IsUsingOverlaySelection()` or `UsesOverlaySelection()` or maybe `HasOverlaySelection()`, but any of these choices are acceptable, so please choose whichever you prefer.
> 
> >

I chooses `UsesOverlaySelection()` and added `DisableOverlaySelection()` as you suggested. also added documentation for them.

I think this is ready now, but as always I didn't tests this under `wxOSX`. so if you could please do it (or anyone interested) before merging.

TIA!

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -151,6 +162,16 @@ class WXDLLIMPEXP_CORE wxGridSelection
+    //
+    // - Simple rectangle (using wxDC::DrawRectangle() if it is empty and the bounding box is valid.
+    // - Simple polygon (using wxDC::DrawPolygon()) if it represents a simple polygon.
+    // - Poly-polygon (using wxDC::DrawPolyPolygon()) if it consists of multi-polygons.
```
Really minor, but I'd say:
```suggestion
    // - Poly-polygon (using wxDC::DrawPolyPolygon()) if it consists of multiple polygons.
```

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -18,6 +18,9 @@
 #include "wx/vector.h"
 
+// Forward declaration
+namespace wxGridPrivate { class PolyPolygon; }
```
I wonder if we should call this class `SelectionShape` instead? This name seems to carry more meaning than the current one (especially because it doesn't even always consists of multiple polygons).

AliKet replied:
Done

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -114,6 +117,11 @@ class WXDLLIMPEXP_CORE wxGridSelection
     void EndSelecting();
     void CancelSelecting();
 
+    // Return the PolyPolygon object. Call ComputePolyPolygon() if necessary.
```
This comment should probably mention that it returns the intersection of the shape of the selection with the given rectangle (if I understand what it does correctly).

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -139,6 +147,9 @@ class WXDLLIMPEXP_CORE wxGridSelection
                          const wxGridBlockCoords& block);
 
+    // Called each time the selection changed or scrolled to recompute m_polyPolygon.
+    void ComputePolyPolygon(bool refreshLabelWindows = false, const wxRect& renderExtent = {});
```
Sorry, I still think that this function shouldn't have `refreshLabelWindows` argument and that the case of `refreshLabelWindows = true` should be handled in the single place where it's used. I realize that this means we're going to potentially call `Refresh()` twice but I don't think this is a problem in practice, the refresh rectangles should be coalesced anyhow.

And it would make this function much easier to understand.

AliKet replied:
Done

AliKet replied:
`ComputePolyPolygon()` has been renamed to `ComputeSelectionShape()`. Because the new name seems strange to me now! I'm thinking of renaming it to `UpdateSelectionShape()` instead, what do you think ?

vadz replied:
Both `Compute` and `Update` seem fine to me, with maybe a slight preference for the latter, but mostly I just think that `SelectionShape()` is much better than `PolyPolygon` here.

Thanks!

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -151,6 +162,16 @@ class WXDLLIMPEXP_CORE wxGridSelection
+    //
+    std::unique_ptr<wxGridPrivate::PolyPolygon> m_polyPolygon;
+
+    size_t                                      m_isAnyLabelHighlighted = 0;
```
I'd also appreciate a comment about the meaning of this one as its name (`isXXX`) implies `bool` type for me, but here it is `size_t`, so clearly there is something else going on here, but I don't really know what.

vadz commented on include/wx/generic/gridsel.h:
```diff
@@ -151,6 +162,16 @@ class WXDLLIMPEXP_CORE wxGridSelection
+    // - Simple polygon (using wxDC::DrawPolygon()) if it represents a simple polygon.
+    // - Poly-polygon (using wxDC::DrawPolyPolygon()) if it consists of multi-polygons.
+    //
+    std::unique_ptr<wxGridPrivate::PolyPolygon> m_polyPolygon;
```
We should `#include <memory>` in this file instead of relying on it being done implicitly in some other header.

AliKet replied:
Done

vadz commented on include/wx/generic/grid.h:
```diff
@@ -2267,6 +2296,12 @@ class WXDLLIMPEXP_CORE wxGrid : public wxScrolledCanvas
 
+    // Return @true if overlay selection can be used (wxUSE_GRAPHICS_CONTEXT=1)
+    // and DisableOverlaySelection() hadn't been called.
+    static bool UsesOverlaySelection();
+
+    static void DisableOverlaySelection();
```
Sorry, but why do these functions have to be global, i.e. class-static? I thought they would be per-grid, so that you could customize the selection of a particular grid without affecting anything else.

Maybe it would make sense to have some global `DisableOverlaySelectionByDefault()` that would set `gs_useOverlaySelection` to `false`, but I'd still add `m_useOverlaySelection`, that would be initialized with `gs_useOverlaySelection` but could be set to `false` using non-static `DisableOverlaySelection()` without affecting any other grids.

AliKet replied:
Done without adding `DisableOverlaySelectionByDefault()`.

vadz commented on include/wx/generic/private/grid.h:
```diff
@@ -1227,6 +1236,11 @@ wxGetContentRect(wxSize contentSize,
                  int hAlign,
                  int vAlign);
 
+inline bool operator<(const wxPoint& pt1, const wxPoint& pt2)
```
I'd prefer to define a custom comparator and pass it to `std::{set,map}` explicitly. Comparing points lexicographically doesn't always make sense and while it's unlikely to affect anything right now, when this is declared in a private header, it still might and it could be very annoying to run into problems due to this later.

AliKet replied:
Done

vadz commented on include/wx/generic/private/grid.h:
```diff
@@ -1280,6 +1294,235 @@ TryGetValueAsDate(wxDateTime& result,
+private:
+    void CalcBoundingBox(const std::vector<wxPoint>& points);
+
+    // m_counts contains the number of points in each of the polygons in m_points.
```
So the data structure here is a vector of points in all polygons stored as a single continuous vector? I.e. if we have 2 polygons P1 and P2, `m_points` contains P1_first, P1_second, ..., P1_last, P2_first, ..., P2_last?

Is this done for efficiency reasons? Because for me a natural approach would be to just have `vector<vector<wxPoint>> m_polygons` and I wonder why is it not being used here?

AliKet replied:
> So the data structure here is a vector of points in all polygons stored as a single continuous vector? I.e. if we have 2 polygons P1 and P2, `m_points` contains P1_first, P1_second, ..., P1_last, P2_first, ..., P2_last?
> 

Yes.

> Is this done for efficiency reasons? Because for me a natural approach would be to just have `vector<vector<wxPoint>> m_polygons` and I wonder why is it not being used here?

Well, the data structure is modeled in such a way that we can use it to draw a poly-polygon using [DrawPolyPolygon()](https://docs.wxwidgets.org/3.2/classwx_d_c.html#ae760da96008debe42e18195d8ac8df3d) in a simple and straightforward way. I don't see how a `vector<vector<wxPoint>>` would simplify things here to be honest?

vadz replied:
Sorry, I forgot about this `DrawPolyPolygon()` convention, I thought the polygons would be drawn one by one, but if we use this function then storing it like this makes sense, thanks, even if using a vector of vectors would still definitely be simpler (because a polygon is a vector of points, and so many polygons is a vector of such vectors).

vadz commented on include/wx/generic/private/grid.h:
```diff
@@ -1280,6 +1294,235 @@ TryGetValueAsDate(wxDateTime& result,
+        m_minY = 0, m_maxY = 0;
+};
+
+// This class is just a helper that simply converts (using the sweep line algorithm)
```
Sorry but why do we need a class for what looks like just a function? I had a lot of trouble understanding what it does, but now that I (think I) understand it, it looks like it could be just a function of `PolyPolygon`, why do we have to make a separate class for it?

vadz reviewed (commented):
Thanks again for all this work, but I still wonder if it can't be simplified a bit more. I'm sorry for all the comments, but wxGrid code is already awfully complicated and so I'd really like to make an effort to reduce additional increase of its complexity as much as possible.

vadz commented:
I've tested this PR under Mac, seems to work great, thanks!

![image](https://github.com/wxWidgets/wxWidgets/assets/146917/dab0b9cb-b43b-4248-b80f-2bcad89a5c9e)

Commit:
Remove refreshLabelWindows parameter from wxGridSelection::ComputePolyPolygon()

This parameter was first added as an optimization to avoid unnecessary Refresh()
calls which we can live with since the cases where we need it are indeed rare and
not too expensive.

Also change m_isAnyLabelHighlighted to a boolean m_updateHighlightedLabels.

No change in behaviour with this commit.

Commit:
Use SelectionShape instead of PolyPolygon to refer to the selection shape

Since selection shape is not clearly expressed using PolyPolygon, SelectionShape
is now used instead to denote class/function/variable entities:

	- PolyPolygon to SelectionShape (class)
 	- GetPolyPolygon() to GetSelectionShape()
 	- ComputePolyPolygon() to ComputeSelectionShape()
 	- InvalidatePolyPolygon() to InvalidateSelectionShape()
 	- m_polyPolygon to m_selectionShape

Commit:
Get rid of PolyPolygonHelper class

Simply replaced by FromRectangles() function added to SelectionShape class.

Commit:
Use custom comparator to compare wxPoints in SelectionShape class

AliKet commented:
@vadz Unrelated to this PR but I just noticed this function [wxGrid::DrawAllGridLines()](https://github.com/wxWidgets/wxWidgets/blob/7f5ca30e3eebbdb85222fbe329a5100b4da65804/include/wx/generic/grid.h#L1639) which is undocumented and not used or referenced anywhere and I'm wondering if it's safe to get rid of it?

vadz commented:
> @vadz Unrelated to this PR but I just noticed this function [wxGrid::DrawAllGridLines()](https://github.com/wxWidgets/wxWidgets/blob/7f5ca30e3eebbdb85222fbe329a5100b4da65804/include/wx/generic/grid.h#L1639) which is undocumented and not used or referenced anywhere and I'm wondering if it's safe to get rid of it?

Yes, it seems pretty useless to me, so I think we can remove it, thanks.

vadz commented:
Just to confirm: this is ok to merge already, right?

AliKet commented:
> Just to confirm: this is ok to merge already, right?

Yes.

vadz commented:
> E.g. when complete row(s) are selected, I don't want all the column headers to be highlighted. For some grids I would like to have the overlay mode for the selected cell(s), but I don't want the headers to be highlighted for the current cell or the selection. For some grids, I would like the current row and col label being highlighted, but only when the grid actually has the focus.

I think the current implementation works as in spreadsheet programs, which means that this is probably the most familiar thing for most users (I've checked LibreOffice Calc. Excel and Google Sheets and they all do it like this).

> Maybe the active cell should be in blue as well if there is no selection. (I find it a bit inconsistent that the row and col header are blue like for a selection, but the cell is not.)
> 
> As of now, I find the implementation being a bit too distracting from other contents and I have to disable it. Traditionally 'blue' means 'focused' to me (I'm on Windows). Maybe, the overlay colour should switch from blue to grey while not focused?

Excel does use grey rather than blue, so maybe this colour should be platform-dependent and grey under Windows...

> I'm not sure what would be the best way:
> 
>     * add arguments to `DisableOverlaySelection`
> 
>     * add style flags?

 I'd rather avoid the use of style flags, but we could add some SetOverlaySelectionPreferences() (better name welcome) taking wxOverlaySelectionPreferences object (ditto) which would provide an extensible way to customize behaviour and colours. I don't plan doing this myself, but if you'd like to make a PR adding this, it would be welcome.

TIA!

DietmarSchwertberger commented:
Thanks for feedback. Currently I have two other projects, but after these I will look into this and experiment a bit.
Is there an example for a Preferences object? Otherwise I would prefer the method plus private properties for easier implementation. I would think that for any grid the preferences are set only once, so I'm not sure whether more abstraction would really add much value.

vadz commented:
I was thinking of a `wxSizerFlags`-like object, i.e. something that could be built using a builder pattern (a.k.a. chained function calls on the same object) and would contain all the customizable properties — this is tidier than having half a dozen or more of separate parameters/variables.
